src/common/version/build.rs

```diff
@@ -19,10 +19,13 @@ use build_data::{format_timestamp, get_source_time};
 use shadow_rs::{CARGO_METADATA, CARGO_TREE};
 
 fn main() -> shadow_rs::SdResult<()> {
-    println!(
-        "cargo:rerun-if-changed={}/.git/refs/heads",
-        env!("CARGO_RUSTC_CURRENT_DIR")
-    );
+    // The "CARGO_WORKSPACE_DIR" is set manually (not by Rust itself) in Cargo config file, to
+    // solve the problem where the "CARGO_MANIFEST_DIR" is not what we want when this repo is
+    // made as a submodule in another repo.
+    let src_path = env::var("CARGO_WORKSPACE_DIR").or_else(|_| env::var("CARGO_MANIFEST_DIR"))?;
+
+    // `CARGO_RUSTC_CURRENT_DIR` is removed, use `src_path` instead.
+    println!("cargo:rerun-if-changed={}/.git/refs/heads", src_path);
```

Comment:
The `CARGO_RUSTC_CURRENT_DIR` is removed due to [this](https://github.com/rust-lang/cargo/issues/14798). However, I haven't found a way to set this ENV correctly. `CARGO_WORKSPACE_DIR` is not set and `CARGO_MANIFEST_DIR` is wrong.

Reply:
According to https://github.com/baoyachi/shadow-rs/issues/175#issuecomment-2543936776 upgrade `shadow-rs` and use `BuildPattern::Lazy` should fix the problem. Let's try and see.